src/imitation/util/logger.py

```diff
@@ -0,0 +1,189 @@
+  os.makedirs(folder, exist_ok=True)
+  output_formats = [sb_logger.make_output_format(f, folder)
+                    for f in format_strs]
+  return output_formats
+
+
+class _HierarchicalLogger(sb_logger.Logger):
+
+  def __init__(self,
+               default_logger: sb_logger.Logger,
+               format_strs: Sequence[str] = ('stdout', 'log', 'csv')):
+    """A logger with a context for accumulating mean values.
+
+    Args:
+      default_logger: The default logger when not in the a `accumulate_means`
+        context. Also the logger to which mean values are written to when
+        contexts are over.
+      format_strs: An list of output format strings that should be used by
+        every Logger initialized by this class during an `AccumulatingMeans`
+        context. For details on available output formats see
+        `stable_baselines.logger.make_output_format`.
+    """
+    self.default_logger = default_logger
+    self.current_logger = None
+    self._cached_loggers = {}
+    self.format_strs = format_strs
+    super().__init__(folder=self.default_logger.dir, output_formats=None)
+
+  @contextlib.contextmanager
+  def accumulate_means(self, subdir: str):
+    """Temporarily modifies this _HierarchicalLogger to accumulate means values.
+
+    During this context, `self.logkv(key, value)` writes the "raw" values in
+    "{self.default_logger.log_dir}/{subdir}" under the key "raw/{subdir}/{key}".
+    At the same time, any call to `self.logkv` will also accumulate mean values
+    on the default logger by calling
+    `self.default_logger.logkv_mean(f"mean/{subdir}/{key}", value)`.
+
+    During the context, `self.logkv(key, value)` will write the "raw" values in
+    `"{self.default_logger.log_dir}/subdir"` under the key "raw/{subdir}/key".
+
+    After the context exits, calling `self.dumpkvs()` will write the means
+    of all the "raw" values accumulated during this context to
+    `self.default_logger` under keys with the prefix `mean/{subdir}/`
+
+    Args:
+      subdir: A string key which determines the `folder` where raw data is
+        written and temporary logging prefixes for raw and mean data. Entering
+        an `accumulate_means` context in the future with the same `subdir`
+        will safely append to logs written in this folder rather than
+        overwrite.
+    """
+    if self.current_logger is not None:
+      raise RuntimeError("Nested `accumulate_means` context")
+
+    if subdir in self._cached_loggers:
+      logger = self._cached_loggers[subdir]
+    else:
+      folder = os.path.join(self.mean_logger.dir, "raw", subdir)
+      os.makedirs(folder)
+      output_formats = _build_output_formats(folder, self.format_strs)
+      logger = sb_logger.Logger(folder, output_formats)
+      self._cached_loggers[subdir] = logger
+
+    try:
+      self.current_logger = logger
+      yield
+    finally:
+      self.current_logger = None
+
+  def logkv(self, key, val):
+    if self.current_logger is not None:
```

Comment:
can't we reuse `self._logger` here to avoid some of the checks?

Reply:
No, because we need to write to two different loggers (it isn't clear from this garbled commit sorry)